WordPress/Classes/Networking/LoginClient.swift

```diff
@@ -78,4 +81,134 @@ actor LoginClient {
             "error": error.localizedDescription
         ])
     }
+
+    @MainActor
+    static func displaySelfHostedLoginView(in navigationController: UINavigationController) {
+        let loginClient = LoginClient(session: .shared)
+        let vm = SelfHostedLoginViewModel(loginClient: loginClient)
+        let loginViewController = UIHostingController(rootView: LoginWithUrlView(viewModel: vm))
+
+        vm.onLoginComplete = {
+            loginViewController.dismiss(animated: true)
+        }
+
+        navigationController.pushViewController(loginViewController, animated: true)
+    }
+}
+
+class SelfHostedLoginViewModel: LoginWithUrlView.ViewModel {
+    private let loginClient: LoginClient
+    private let presentationContextProvider = SelfHostedLoginViewModelAuthenticationContextProvider()
+
+    var onLoginComplete: (() -> Void)?
+
+    init(loginClient: LoginClient) {
+        self.loginClient = loginClient
+        super.init()
+    }
+
+    override func startLogin() {
+        debugPrint("Attempting login with \(super.urlField)")
+        Task {
+            do {
+                let authUrl = try await self.buildLoginUrl(from: super.urlField)
+                debugPrint("Presenting Login Page \(authUrl)")
+
+                let urlWithToken = try await withUnsafeThrowingContinuation { continuation in
+                    let session = ASWebAuthenticationSession(url: authUrl, callbackURLScheme: "x-wpcom-login") { url, error in
+
+                        if let url {
+                            continuation.resume(returning: url)
+                        }
+
+                        if let error {
+                            continuation.resume(throwing: error)
+                        }
+                    }
+                    session.presentationContextProvider = self.presentationContextProvider
+                    session.start()
+                }
+
+                let credentials = try SelfHostedLoginDetails.fromApplicationPasswordResponse(urlWithToken)
+                let blogOptions = try await loadSiteOptions(details: credentials)
+
+                // Only store the new site after credentials are validated.
+                let _ = try await Blog.createRestApiBlog(with: credentials, in: ContextManager.shared)
+
+                // Automatically show the new site.
+                assert(WordPressAuthenticator.shared.delegate != nil)
+                let wporg = WordPressOrgCredentials(
+                    username: credentials.username,
+                    password: credentials.password,
+                    xmlrpc: credentials.derivedXMLRPCRoot.absoluteString,
+                    options: blogOptions
+                )
+                WordPressAuthenticator.shared.delegate!.sync(credentials: .init(wporg: wporg)) {
```

Comment:
I would suggest implementing a simple `LoginViewController` and use `ASWebAuthenticationSession` and `WordPressAuthenticator.shared.delegate` directly from it.  UIKit will probably be a better option too. Let's not pretend it is a ViewModel. It makes it harder to follow this code.

Reply:
I added some code here to make the end-to-end flow work. I plan to refactor this part in a separate PR.